File: mobile/lib/utils/validation.dart
import 'dart:core';

/// Common validation utilities for the mobile app
class ValidationUtils {
  // Email validation regex
  static final RegExp _emailRegex = RegExp(
    r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$',
  );

  // Strong password regex components
  static final RegExp _uppercaseRegex = RegExp(r'[A-Z]');
  static final RegExp _lowercaseRegex = RegExp(r'[a-z]');
  static final RegExp _digitRegex = RegExp(r'\d');
  static final RegExp _specialCharRegex = RegExp(r'[!@#$%^&*()_+\-=\[\]{};\':"\\|,.<>/?~`]');

  /// Validates if email format is correct
  static bool isValidEmail(String email) {
    if (email.isEmpty) return false;
    return _emailRegex.hasMatch(email.trim());
  }

  /// Validates password strength
  /// Returns null if valid, error message if invalid
  static String? validatePassword(String password) {
    if (password.isEmpty) {
      return 'Password is required';
    }

    if (password.length < 8) {
      return 'Password must be at least 8 characters';
    }

    if (password.length > 128) {
      return 'Password must be at most 128 characters';
    }

    if (!_uppercaseRegex.hasMatch(password)) {
      return 'Password must contain at least one uppercase letter';
    }

    if (!_lowercaseRegex.hasMatch(password)) {
      return 'Password must contain at least one lowercase letter';
    }

    if (!_digitRegex.hasMatch(password)) {
      return 'Password must contain at least one digit';
    }

    if (!_specialCharRegex.hasMatch(password)) {
      return 'Password must contain at least one special character';
    }

    return null; // Password is valid
  }

  /// Validates email format
  /// Returns null if valid, error message if invalid
  static String? validateEmail(String email) {
    if (email.isEmpty) {
      return 'Email is required';
    }

    if (!isValidEmail(email)) {
      return 'Please enter a valid email address';
    }

    return null;
  }

  /// Validates required string field
  static String? validateRequired(String? value, String fieldName) {
    if (value == null || value.trim().isEmpty) {
      return '$fieldName is required';
    }
    return null;
  }

  /// Validates string length
  static String? validateLength(String? value, String fieldName, {int? min, int? max}) {
    if (value == null) return null;

    if (min != null && value.length < min) {
      return '$fieldName must be at least $min characters';
    }

    if (max != null && value.length > max) {
      return '$fieldName must be at most $max characters';
    }

    return null;
  }

  /// Validates that two strings match (e.g., password confirmation)
  static String? validateMatch(String? value1, String? value2, String fieldName) {
    if (value1 != value2) {
      return '$fieldName do not match';
    }
    return null;
  }

  /// Validates positive number
  static String? validatePositiveNumber(String? value, String fieldName) {
    if (value == null || value.isEmpty) {
      return '$fieldName is required';
    }

    final number = int.tryParse(value);
    if (number == null || number <= 0) {
      return '$fieldName must be a positive number';
    }

    return null;
  }

  /// Validates number range
  static String? validateNumberRange(String? value, String fieldName, {int? min, int? max}) {
    if (value == null || value.isEmpty) {
      return '$fieldName is required';
    }

    final number = int.tryParse(value);
    if (number == null) {
      return '$fieldName must be a valid number';
    }

    if (min != null && number < min) {
      return '$fieldName must be at least $min';
    }

    if (max != null && number > max) {
      return '$fieldName must be at most $max';
    }

    return null;
  }

  /// Validates that a list is not empty
  static String? validateListNotEmpty<T>(List<T>? list, String fieldName) {
    if (list == null || list.isEmpty) {
      return '$fieldName cannot be empty';
    }
    return null;
  }

  /// Validates URL format
  static String? validateUrl(String? value) {
    if (value == null || value.isEmpty) return null;

    const pattern = r'^https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_\+.~#?&//=]*)$';
    final regex = RegExp(pattern);

    if (!regex.hasMatch(value)) {
      return 'Please enter a valid URL';
    }

    return null;
  }

  /// Sanitizes string input by trimming and removing control characters
  static String sanitizeString(String input) {
    return input.trim().replaceAll(RegExp(r'[\x00-\x1f\x7f]'), '');
  }

  /// Validates and sanitizes string input
  static String? validateAndSanitizeString(
    String? value,
    String fieldName, {
    bool required = true,
    int? minLength,
    int? maxLength,
  }) {
    if (value == null || value.isEmpty) {
      return required ? '$fieldName is required' : null;
    }

    final sanitized = sanitizeString(value);

    if (required && sanitized.isEmpty) {
      return '$fieldName is required';
    }

    return validateLength(sanitized, fieldName, min: minLength, max: maxLength);
  }
}

/// Validation result that can contain multiple errors
class ValidationResult {
  final Map<String, String> errors;

  ValidationResult() : errors = {};

  ValidationResult.fromErrors(this.errors);

  /// Returns true if validation passed (no errors)
  bool get isValid => errors.isEmpty;

  /// Returns true if validation failed (has errors)
  bool get hasErrors => errors.isNotEmpty;

  /// Adds an error for a specific field
  void addError(String field, String message) {
    errors[field] = message;
  }

  /// Gets error message for a specific field
  String? getError(String field) {
    return errors[field];
  }

  /// Gets all error messages as a list
  List<String> get allErrors => errors.values.toList();

  /// Gets all errors as a formatted string
  String get errorsAsString => errors.values.join('\n');
}

/// Form validation helper that accumulates validation errors
class FormValidator {
  final ValidationResult _result = ValidationResult();

  /// Validates a field and adds any errors to the result
  FormValidator validateField(String field, String? Function() validator) {
    final error = validator();
    if (error != null) {
      _result.addError(field, error);
    }
    return this;
  }

  /// Validates email field
  FormValidator validateEmailField(String field, String? value) {
    return validateField(field, () => ValidationUtils.validateEmail(value ?? ''));
  }

  /// Validates password field
  FormValidator validatePasswordField(String field, String? value) {
    return validateField(field, () => ValidationUtils.validatePassword(value ?? ''));
  }

  /// Validates required field
  FormValidator validateRequiredField(String field, String? value) {
    return validateField(field, () => ValidationUtils.validateRequired(value, field));
  }

  /// Validates password confirmation
  FormValidator validatePasswordConfirmation(String password, String confirmation) {
    return validateField('confirmPassword', () {
      if (confirmation.isEmpty) {
        return 'Password confirmation is required';
      }
      return ValidationUtils.validateMatch(password, confirmation, 'Passwords');
    });
  }

  /// Gets the validation result
  ValidationResult get result => _result;

  /// Returns true if all validations passed
  bool get isValid => _result.isValid;

  /// Returns true if any validations failed
  bool get hasErrors => _result.hasErrors;
}

/// Common form field validators
class FormFieldValidators {
  /// Email field validator
  static String? email(String? value) {
    return ValidationUtils.validateEmail(value ?? '');
  }

  /// Password field validator
  static String? password(String? value) {
    return ValidationUtils.validatePassword(value ?? '');
  }

  /// Required field validator
  static String? Function(String?) required(String fieldName) {
    return (String? value) => ValidationUtils.validateRequired(value, fieldName);
  }

  /// Name field validator (required, 2-50 characters)
  static String? name(String? value) {
    final required = ValidationUtils.validateRequired(value, 'Name');
    if (required != null) return required;

    return ValidationUtils.validateLength(value, 'Name', min: 2, max: 50);
  }

  /// Task title validator (required, 1-200 characters)
  static String? taskTitle(String? value) {
    final required = ValidationUtils.validateRequired(value, 'Task title');
    if (required != null) return required;

    return ValidationUtils.validateLength(value, 'Task title', min: 1, max: 200);
  }

  /// Task description validator (optional, max 1000 characters)
  static String? taskDescription(String? value) {
    if (value == null || value.isEmpty) return null;
    return ValidationUtils.validateLength(value, 'Description', max: 1000);
  }

  /// Pomodoro duration validator (1-480 minutes)
  static String? pomodoroDuration(String? value) {
    return ValidationUtils.validateNumberRange(value, 'Duration', min: 1, max: 480);
  }

  /// Break duration validator (1-60 minutes)
  static String? breakDuration(String? value) {
    return ValidationUtils.validateNumberRange(value, 'Break duration', min: 1, max: 60);
  }
}

File: specs/004-review/tracking/backend-backup/internal/utils/validation.go
package utils

import (
	"fmt"
	"regexp"
	"strings"
	"time"

	"github.com/google/uuid"
)

// ValidationError represents a validation error with field and message
type ValidationError struct {
	Field   string `json:"field"`
	Message string `json:"message"`
}

func (e ValidationError) Error() string {
	return fmt.Sprintf("%s: %s", e.Field, e.Message)
}

// ValidationErrors represents multiple validation errors
type ValidationErrors []ValidationError

func (e ValidationErrors) Error() string {
	if len(e) == 0 {
		return "validation failed"
	}

	messages := make([]string, len(e))
	for i, err := range e {
		messages[i] = err.Error()
	}
	return strings.Join(messages, "; ")
}

// Validator provides common validation functions
type Validator struct {
	errors ValidationErrors
}

// NewValidator creates a new validator instance
func NewValidator() *Validator {
	return &Validator{
		errors: make(ValidationErrors, 0),
	}
}

// AddError adds a validation error
func (v *Validator) AddError(field, message string) {
	v.errors = append(v.errors, ValidationError{
		Field:   field,
		Message: message,
	})
}

// AddErrorf adds a formatted validation error
func (v *Validator) AddErrorf(field, format string, args ...interface{}) {
	v.AddError(field, fmt.Sprintf(format, args...))
}

// HasErrors returns true if there are validation errors
func (v *Validator) HasErrors() bool {
	return len(v.errors) > 0
}

// Errors returns all validation errors
func (v *Validator) Errors() ValidationErrors {
	return v.errors
}

// Error implements the error interface
func (v *Validator) Error() error {
	if len(v.errors) == 0 {
		return nil
	}
	return v.errors
}

// Common validation functions

// ValidateRequired validates that a string field is not empty
func (v *Validator) ValidateRequired(field, value string) {
	if strings.TrimSpace(value) == "" {
		v.AddError(field, "is required")
	}
}

// ValidateMinLength validates minimum string length
func (v *Validator) ValidateMinLength(field, value string, minLength int) {
	if len(value) < minLength {
		v.AddErrorf(field, "must be at least %d characters", minLength)
	}
}

// ValidateMaxLength validates maximum string length
func (v *Validator) ValidateMaxLength(field, value string, maxLength int) {
	if len(value) > maxLength {
		v.AddErrorf(field, "must be at most %d characters", maxLength)
	}
}

// ValidateLength validates string length range
func (v *Validator) ValidateLength(field, value string, minLength, maxLength int) {
	v.ValidateMinLength(field, value, minLength)
	v.ValidateMaxLength(field, value, maxLength)
}

// ValidateEmail validates email format
func (v *Validator) ValidateEmail(field, email string) {
	emailRegex := regexp.MustCompile(`^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`)
	if !emailRegex.MatchString(email) {
		v.AddError(field, "must be a valid email address")
	}
}

// ValidateUUID validates UUID format
func (v *Validator) ValidateUUID(field, id string) {
	if _, err := uuid.Parse(id); err != nil {
		v.AddError(field, "must be a valid UUID")
	}
}

// ValidatePositiveInt validates that an integer is positive
func (v *Validator) ValidatePositiveInt(field string, value int) {
	if value <= 0 {
		v.AddError(field, "must be positive")
	}
}

// ValidateRange validates that an integer is within a range
func (v *Validator) ValidateRange(field string, value, min, max int) {
	if value < min || value > max {
		v.AddErrorf(field, "must be between %d and %d", min, max)
	}
}

// ValidateIn validates that a string is in a list of allowed values
func (v *Validator) ValidateIn(field, value string, allowed []string) {
	for _, allowedValue := range allowed {
		if value == allowedValue {
			return
		}
	}
	v.AddErrorf(field, "must be one of: %s", strings.Join(allowed, ", "))
}

// ValidateURL validates URL format
func (v *Validator) ValidateURL(field, url string) {
	urlRegex := regexp.MustCompile(`^https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_\+.~#?&//=]*)$`)
	if !urlRegex.MatchString(url) {
		v.AddError(field, "must be a valid URL")
	}
}

// ValidateDate validates that a time is not zero and optionally in the future
func (v *Validator) ValidateDate(field string, date time.Time, mustBeFuture bool) {
	if date.IsZero() {
		v.AddError(field, "is required")
		return
	}

	if mustBeFuture && date.Before(time.Now()) {
		v.AddError(field, "must be in the future")
	}
}

// ValidatePassword validates password strength
func (v *Validator) ValidatePassword(field, password string) {
	if len(password) < 8 {
		v.AddError(field, "must be at least 8 characters")
		return
	}

	if len(password) > 128 {
		v.AddError(field, "must be at most 128 characters")
		return
	}

	hasUpper := regexp.MustCompile(`[A-Z]`).MatchString(password)
	hasLower := regexp.MustCompile(`[a-z]`).MatchString(password)
	hasDigit := regexp.MustCompile(`\d`).MatchString(password)
	hasSpecial := regexp.MustCompile(`[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?~` + "`" + `]`).MatchString(password)

	if !hasUpper {
		v.AddError(field, "must contain at least one uppercase letter")
	}
	if !hasLower {
		v.AddError(field, "must contain at least one lowercase letter")
	}
	if !hasDigit {
		v.AddError(field, "must contain at least one digit")
	}
	if !hasSpecial {
		v.AddError(field, "must contain at least one special character")
	}
}

// ValidateArrayNotEmpty validates that an array is not empty
func (v *Validator) ValidateArrayNotEmpty(field string, arr interface{}) {
	switch a := arr.(type) {
	case []string:
		if len(a) == 0 {
			v.AddError(field, "cannot be empty")
		}
	case []int:
		if len(a) == 0 {
			v.AddError(field, "cannot be empty")
		}
	default:
		v.AddError(field, "invalid array type")
	}
}

// ValidateArrayMaxLength validates maximum array length
func (v *Validator) ValidateArrayMaxLength(field string, arr []string, maxLength int) {
	if len(arr) > maxLength {
		v.AddErrorf(field, "cannot have more than %d items", maxLength)
	}
}

// ValidateConditional validates a field only if a condition is true
func (v *Validator) ValidateConditional(condition bool, validationFunc func()) {
	if condition {
		validationFunc()
	}
}

// Utility functions for common validations

// IsValidEmail checks if an email is valid
func IsValidEmail(email string) bool {
	emailRegex := regexp.MustCompile(`^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`)
	return emailRegex.MatchString(email)
}

// IsValidUUID checks if a string is a valid UUID
func IsValidUUID(id string) bool {
	_, err := uuid.Parse(id)
	return err == nil
}

// IsValidPassword checks if a password meets strength requirements
func IsValidPassword(password string) bool {
	if len(password) < 8 || len(password) > 128 {
		return false
	}

	hasUpper := regexp.MustCompile(`[A-Z]`).MatchString(password)
	hasLower := regexp.MustCompile(`[a-z]`).MatchString(password)
	hasDigit := regexp.MustCompile(`\d`).MatchString(password)
	hasSpecial := regexp.MustCompile(`[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?~` + "`" + `]`).MatchString(password)

	return hasUpper && hasLower && hasDigit && hasSpecial
}

// SanitizeString removes potentially harmful characters from a string
func SanitizeString(input string) string {
	// Remove null bytes and control characters
	sanitized := regexp.MustCompile(`[\x00-\x1f\x7f]`).ReplaceAllString(input, "")
	// Trim whitespace
	return strings.TrimSpace(sanitized)
}

// ValidateAndSanitizeString validates and sanitizes a string field
func (v *Validator) ValidateAndSanitizeString(field string, value *string, required bool, minLen, maxLen int) {
	if value == nil {
		if required {
			v.AddError(field, "is required")
		}
		return
	}

	*value = SanitizeString(*value)

	if required {
		v.ValidateRequired(field, *value)
	}

	if *value != "" {
		v.ValidateLength(field, *value, minLen, maxLen)
	}
}